src/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -1125,6 +1123,26 @@ internal void Unregister()
                 _handlerRegistered = false;
                 Interop.Sys.UnregisterForCtrl();
             }
+
+            private static void OnBreakEvent(Interop.Sys.CtrlCode ctrlCode)
+            {
+                // This is called on the native signal handling thread. We need to move to another thread so signal handling is
+                // not blocked. Otherwise we may get deadlocked when the handler depends on work triggered from the signal handling thread.
+                // We spin up a new thread so the break event is handled promptly.
+                Thread handlerThread = new Thread(HandleBreakEvent) { IsBackground = true };
+                handlerThread.Start(ctrlCode);
+            }
+
+            private static void HandleBreakEvent(object state)
+            {
+                var ctrlCode = (Interop.Sys.CtrlCode)state;
+                ConsoleSpecialKey controlKey = (ctrlCode == Interop.Sys.CtrlCode.Break ? ConsoleSpecialKey.ControlBreak : ConsoleSpecialKey.ControlC);
+                bool cancel = Console.HandleBreakEvent(controlKey);
+                if (!cancel)
+                {
+                    Interop.Sys.RestoreAndHandleCtrl(ctrlCode);
```

Comment:
It should be super rare, tbut it's now possible this could be invoked multiple times concurrently.  Would anything go really wrong if that happened?

Reply:
All functions we call in `RestoreAndHandleCtrl` are [async-signal-safe functions](http://man7.org/linux/man-pages/man7/signal-safety.7.html). So these functions are safe to call concurrently.

The docs for `CancelKeyPress`: _The event handler for this event is executed on a thread pool thread._ So the user is responsible for making the CancelKeyPress handler thread-safe.